core/domain/exp_domain.py

```diff
@@ -2989,6 +2989,90 @@ def _convert_states_v39_dict_to_v40_dict(cls, states_dict):
 
         return states_dict
 
+    @classmethod
+    def _convert_states_v40_dict_to_v41_dict(cls, states_dict):
+        """Converts from version 40 to 41. Version 41 adds
+        TranslatableSetOfUnicodeString and TranslatableSetOfNormalizedString
+        objects to RuleSpec domain objects to allow for translations.
+
+        Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        class ContentIdCounter(python_utils.OBJECT):
+            """This helper class is used to keep track of
+            next_content_id_index and new_content_ids, and provides a
+            function to generate new content_ids.
+            """
+
+            def __init__(self, next_content_id_index):
+                """Initializes a ContentIdCounter object.
+
+                Args:
+                    next_content_id_index: int. The next content id index.
+                """
+                self.new_content_ids = []
+                self.next_content_id_index = next_content_id_index
+
+            def generate_content_id(self, content_id_prefix):
+                """Generate a new content_id from the prefix provided and
+                the next content id index.
+
+                Args:
+                    content_id_prefix: str. The prefix of the content_id.
+
+                Returns:
+                    str. The generated content_id.
+                """
+                content_id = '%s%i' % (
+                    content_id_prefix,
+                    self.next_content_id_index)
+                self.next_content_id_index += 1
+                self.new_content_ids.append(content_id)
+                return content_id
+
+        for state_dict in states_dict.values():
+            # As of Jan 2021, which is when this migration is to be run, only
+            # TextInput and SetInput have translatable rule inputs.
```

Comment:
What about ItemSelectionInput & Drag&Drop? Don't those also have string inputs that would require translating?

Reply:
Ah, to clarify my understanding: the difference for those interactions is that the rule inputs need to be managed such that they use the _same_ content ID as the ones defined in the customization args, whereas for these inputs we have separate content IDs computed as part of the inputs themselves? Is that correct?

Reply:
@shavavo can you confirm the above?

Reply:
@BenHenning that's right- sorry for missing this

Reply:
Thanks for confirming!